src/btree/bt_read.c

```diff
@@ -536,24 +541,24 @@ __page_read(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags)
 	case WT_REF_LIMBO:
 	case WT_REF_LOOKASIDE:
 		WT_ERR(__page_read_lookaside(
-		    session, ref, previous_state, &final_state));
+		    session, ref, previous_state, &final_state, &prepared));
 		break;
 	}
 
 	/*
 	 * Once the page is instantiated, we no longer need the history in
 	 * lookaside.  We leave the lookaside sweep thread to do most cleanup,
-	 * but it can only remove keys that skew newest (if there are entries
-	 * in the lookaside newer than the page, they need to be read back into
-	 * cache or they will be lost).
+	 * but it can only remove committed updates and keys that skew newest
+	 * (if there are entries in the lookaside newer than the page, they need
+	 * to be read back into cache or they will be lost).
 	 *
 	 * There is no reason for the lookaside remove should fail, but ignore
 	 * it if for some reason it fails, we've got a valid page.
 	 *
 	 * Don't free WT_REF.page_las, there may be concurrent readers.
 	 */
 	if (final_state == WT_REF_MEM &&
-	    ref->page_las != NULL && !ref->page_las->skew_newest)
+	    ref->page_las != NULL && (prepared || !ref->page_las->skew_newest))
 		WT_IGNORE_RET(__wt_las_remove_block(
```

Comment:
I figured out why we have entries that aren't being removed from lookaside, because we are ignoring failure of __wt_las_remove_block(). When lookaside sweep or eviction to lookaside has taken a lock, removal from lookaside here fails to obtain the lock and failure gets ignored and we never remove these entries.

To confirm the above theory, I make this function wait on lock (pass true for the last argument), my test passes and there are no prepared entries in lookaside at shutdown. Statistics for removal and insertion also seem to match.

Reply:
@agorrod I am not sure if WT_IGNORE_RET(..) here is okay to do. I am assuming it was being done here because it isn't safe to bail out of this function anymore ?

Reply:
I don't think it's OK for that to fail any longer - we are relying on those removes succeeding now.